exporter/sumologicexporter/sender.go

```
@@ -0,0 +1,277 @@
+		if err != nil {
+			errs = append(errs, err)
+			if sent {
+				droppedRecords = append(droppedRecords, currentRecords...)
+			}
+
+			if !appended {
+				droppedRecords = append(droppedRecords, record)
+			}
+		}
+
+		// If data was sent, cleanup the currentRecords
+		if sent {
+			currentRecords = currentRecords[:0]
+		}
+
+		// If log has been appended to body, add it to the currentRecords
+		if appended {
+			currentRecords = append(currentRecords, record)
+		}
+	}
+
+	if err := s.send(LogsPipeline, strings.NewReader(body.String()), flds); err != nil {
+		errs = append(errs, err)
+		droppedRecords = append(droppedRecords, currentRecords...)
+	}
+
+	if len(errs) > 0 {
+		return droppedRecords, componenterror.CombineErrors(errs)
+	}
+	return droppedRecords, nil
+}
+
+// appendAndSend appends line to the request body that will be sent and sends
+// the accumulated data if the internal buffer has been filled (with maxBufferSize elements).
+// It returns 2 booleans saying whether the data was sent and whether the provided data
+// was appended and an error if anything failed.
+func (s *sender) appendAndSend(
+	line string,
+	pipeline PipelineType,
+	body *strings.Builder,
+	flds fields,
+) (sent bool, appended bool, _ error) {
+	var errors []error
+	// sent gives information if the data was sent or not
+	sent = false
+	// appended keeps state of appending new log line to the body
+	appended = true
+
+	if body.Len() > 0 && body.Len()+len(line) > s.config.MaxRequestBodySize {
+		sent = true
+		if err := s.send(pipeline, strings.NewReader(body.String()), flds); err != nil {
+			errors = append(errors, err)
+		}
+		body.Reset()
+	}
+
+	if body.Len() > 0 {
+		// Do not add newline if the body is empty
+		if _, err := body.WriteString("\n"); err != nil {
+			errors = append(errors, err)
+			appended = false
+		}
+	}
+
+	if _, err := body.WriteString(line); err != nil && appended {
```

This is a bit confusing: if there was a CRLF added in the previous step, `appended` will be false. If an error occurred while trying to write the line to the body, the error will then be swallowed. Is that the intention?

---

nope, logic fixed

---

if adding a CRLF failed, log shouldn't be appended to body